amethyst_tiles/src/morton/mod.rs

```diff
@@ -0,0 +1,232 @@
+
+    answer = answer << 48 | u64::from(MORTON256_Z[((z >> 8) & 0xFF) as usize]) | // shifting second byte
+        u64::from(MORTON256_Y[((y >> 8) & 0x0000_00FF) as usize])|
+        u64::from(MORTON256_X[((x >> 8) & 0x0000_00FF) as usize]);
+
+    (answer << 24 |
+        u64::from(MORTON256_Z[((z) & 0x0000_00FF) as usize]) | // first byte
+        u64::from(MORTON256_Y[((y) & 0x0000_00FF) as usize]) |
+        u64::from(MORTON256_X[((x) & 0x0000_00FF) as usize])) as u32
+}
+
+#[inline]
+pub fn morton_decode_lut(morton: u32) -> (u32, u32, u32) {
+    let single_coord = |morton, shift, table: &[u8]| -> u32 {
+        let mut a: u32 = 0;
+        for i in 0 as u32..4 {
+            a |= u32::from(table[(morton >> ((i * 9) + shift) & 0x0000_01FF) as usize]) << (3 * i);
+        }
+
+        a
+    };
+
+    (
+        single_coord(morton, 0, MORTON512_DECODE_X),
+        single_coord(morton, 0, MORTON512_DECODE_Y),
+        single_coord(morton, 0, MORTON512_DECODE_Z),
+    )
+}
+
+#[inline]
+pub fn morton_encode_intr_3d(x: u32, y: u32, z: u32) -> u32 {
+    use bitintr::Pdep;
+    z.pdep(0x2492_4924) | y.pdep(0x1249_2492) | x.pdep(0x0924_9249)
+}
+
+#[inline]
+pub fn morton_decode_intr_3d(morton: u32) -> (u32, u32, u32) {
+    use bitintr::Pext;
+
+    (
+        morton.pext(0x0924_9249),
+        morton.pext(0x1249_2492),
+        morton.pext(0x2492_4924),
+    )
+}
+
+#[inline]
+//#[cfg(target_feature = "bmi2")]
+pub fn encode(x: u32, y: u32, z: u32) -> u32 {
+    morton_encode_intr_3d(x, y, z)
+}
+
+#[inline]
+//#[cfg(target_feature = "bmi2")]
+pub fn decode(morton: u32) -> (u32, u32, u32) {
+    morton_decode_intr_3d(morton)
+}
+
+#[derive(Default)]
+pub struct Encoder;
+impl CoordinateEncoder for Encoder {
+    fn from_dimensions(_: u32, _: u32, _: u32) -> Self {
+        Self {}
+    }
+
+    #[inline]
+    fn encode(&self, x: u32, y: u32, z: u32) -> Option<u32> {
+        Some(encode(x, y, z))
+    }
+    #[inline]
+    fn decode(&self, morton: u32) -> Option<(u32, u32, u32)> {
+        Some(decode(morton))
+    }
+}
+
+#[derive(Default)]
+pub struct Encoder2D {
+    dimensions: (u32, u32, u32),
+}
+impl CoordinateEncoder for Encoder2D {
+    fn from_dimensions(x: u32, y: u32, z: u32) -> Self {
+        Self {
+            dimensions: (x, y, z),
+        }
+    }
+
+    #[inline]
+    fn encode(&self, x: u32, y: u32, z: u32) -> Option<u32> {
+        use bitintr::Pdep;
+        let len = self.dimensions.0 * self.dimensions.1;
+        let morton = (x.pdep(0x5555_5555) | y.pdep(0xAAAA_AAAA)) + (z * len);
+
+        Some(morton)
+    }
+    #[inline]
+    fn decode(&self, mut morton: u32) -> Option<(u32, u32, u32)> {
+        use bitintr::Pext;
+
+        let d = self.dimensions.0 * self.dimensions.1;
+
+        let z = morton / d;
+        morton -= z * d;
+
+        Some((morton.pext(0x5555_5555), morton.pext(0xAAAA_AAAA), z))
+    }
+}
+
+/*
+#[inline]
+#[cfg(not(target_feature = "bmi2"))]
```

Comment:
I'm not exatly sure what happens with `bmi2` here. Do you consider it widespread enough to assume? Also, what about non-intel targets (we have to support mobile/wasm)? Or is `Pext` just emulated on non-supporting platforms?

Reply:
It falls back on a standard implementation in the case that bmi2 doesn't exist. I'd like to still at least use the optimized LUT tables though.